File: Shells/Cpp/cspSolver/Variable.cpp
#include"Variable.hpp"

int Variable::namingCounter = 1;


Variable::Variable(std::vector<int> possible_Values, int row, int col, int block) : domain(possible_Values){
    if (domain.size() == 1){
        modified = true;
        unchangeable = true;
    } 
    r = row;
    c = col;
    b = block;
    name = "v" + std::to_string(namingCounter++);
    oldSize = size();
}

Variable& Variable::operator=(const Variable& other){
	r = other.row();
	c = other.col();
	b = other.block();
	modified = other.isModified();
	name = other.getName();
	domain = other.getDomain();
	oldSize = other.oldSize;
	return *this;
}

void Variable::modify(const Variable& other){
	r = other.row();
	c = other.col();
	b = other.block();
	modified = other.isModified();
	name = other.getName();
	domain = other.getDomain();
	oldSize = other.oldSize;
}

Variable::Variable(const Variable& v) : domain(v.domain){
    r = v.row();
    c = v.col();
    b = v.block();
    modified = v.isModified();
    name = v.getName();
}

bool Variable::operator==(const Variable &other) const{
    return (row() == other.row()) && (col() == other.col()) && (block() == other.block());
}

bool Variable::operator!=(const Variable &other) const{
    return !(*this == other);
}

bool Variable::isChangeable() const{
    return domain.size() != 1;
}

bool Variable::isAssigned() const{
    return size() == 1 ? true : false;
}

bool Variable::isModified() const{
    return modified;
}

void Variable::setModified(bool modified){
    this->modified = modified;
    this->domain.setModified(modified);
}

int Variable::row() const{
    return r;
}

int Variable::col() const{
    return c;
}

int Variable::block() const{
    return b;
}

int Variable::getAssignment() const{
    if (!isAssigned()){
        return 0;
    }
    else{
		if(domain.getValues().size() != 1)
			std::cout<<"ERROR!"<<std::endl;
        return domain.getValues()[0];
    }
}

std::vector<int> Variable::Values() const{
    return domain.getValues();
}

int Variable::size() const{
    return domain.size();
}

Domain Variable::getDomain() const{
    return domain;
}

std::string Variable::getName() const{
    return name;
}

void Variable::assignValue(int val){
    setDomain(Domain(val));
}

void Variable::updateDomain(Domain d){
    methodModifiesDomain();
    
    domain = d;
//	std::cout<<size()<<std::endl;
//	std::cout<<isAssigned()<<std::endl;
    modified = true;
}

void Variable::setDomain(Domain d){
    domain = d;
    modified = true;
}

void Variable::removeValueFromDomain(int val) {
    methodModifiesDomain();
    domain.remove(val);
    modified=domain.isModified();
}

void Variable::methodModifiesDomain(){
    int newSize = size();
    if (oldSize > newSize){
        oldSize = newSize;
    }
    //**********
    //trail.push(this);
}

Domain::iterator Variable::begin(){
    return domain.begin();
}

Domain::iterator Variable::end(){
    return domain.end();
}

std::string Variable::toString(){
    //prints node stats
    std::stringstream ss;
    //StringBuilder sb = StringBuilder();
    std::string sep = "";
    ss<<" Name: "<<name;
    ss<<"\tdomain: {";
    //		sb.append(" Name: "+name);
    //		sb.append("\tdomain: {");
    for(int i = 0; i < domain.getValues().size(); ++i){
        ss<<sep<<domain.getValues()[i];
        sep = ",";
        //sb.append(i + ",");
    }
    ss<<"}";
    return ss.str();
}


Domain::Domain(int value){
    values.push_back(value);
}

Domain& Domain::operator=(const Domain& other){
	values = other.getValues();
	modified = other.isModified();
	return *this;
}
//Domain::Domain(int* values){
//    for(int i = 0; i < values.length; i++){
//        this.values.add(values[i]);
//    }
//}

Domain::Domain(std::vector<int> vals)
{
	values = vals;
}

Domain::Domain(const Domain& d){
    for(int i = 0; i < d.getValues().size(); ++i)
        values.push_back(d.getValues()[i]);
}

std::vector<int> Domain::getValues() const{
    return values;
}

bool Domain::contains(int value) const{
    if(std::find(values.begin(), values.end(), value) != values.end())
	    return true;
    return false;
}

int Domain::size() const{
    return values.size();
}

bool Domain::isEmpty() const{
    return values.empty();
}

/**
 * Returns whether or not the domain has been modified.
 *
 * @return true if the domain has been modified
 */
bool Domain::isModified() const{
    return modified;
}


bool Domain::remove(int value){
    int v = value;
    
    if(std::find(values.begin(), values.end(), value) == values.end())
    	return false;
    
    setModified(true);
    values.erase(std::remove(values.begin(), values.end(), value), values.end());
    return true;
}

void Domain::setModified(bool modified){
    this->modified = modified;
}

bool Domain::equals(Domain d) const{
    return false;
}

std::vector<int>::iterator Domain::begin(){
    return values.begin();
}

std::vector<int>::iterator Domain::end(){
    return values.end();
}

std::string Domain::toString(){
    std::stringstream ss;//new StringBuilder("{");
    ss<<"{";
    std::string delim = "";
    for (int i = 0; i < values.size(); ++i){
        ss<<delim<<values[i];
        delim = ",";
    }
    ss<<"}";
    return ss.str();
}

File: Shells/Java/src/cspSolver/Variable.java
package cspSolver;

import java.util.Iterator;
import java.util.List;

public class Variable implements Iterable<Integer> 
{
	//===============================================================================
	// Properties
	//===============================================================================
	
	private static int namingCounter = 1;
	private Domain domain;	
	private int row, col, block;
	private boolean modified;
	private boolean unchangeable;
	private String name;
	private static Trail trail = Trail.getTrail();
	
	private int oldSize;
	
	//===============================================================================
	// Constructors
	//===============================================================================
	
	public Variable(List<Integer> possible_Values, int row, int col, int block)
	{	
		this.domain = new Domain(possible_Values);
		if (size() == 1)
		{
			modified = true;
			unchangeable = true;
		}
		this.row = row;
		this.col = col;
		this.block = block;
		this.name = "v" + namingCounter++;
		this.oldSize = size();
	}
	
	public Variable(Variable v)
	{
		this.domain = v.domain;
		this.row = v.row;
		this.col = v.col;
		this.block = v.block;
		this.modified = v.modified;
		this.name = v.name;
	}
	
	//===============================================================================
	// Accessors
	//===============================================================================
	
	/**
	 * Used for local search. Used to differentiate between intialized assignments
	 * and user assignments.
	 * @return false if this variable was assigned at initialization.
	 */
	public boolean isChangeable()
	{
		return !unchangeable;
	}
	
	public boolean isAssigned()
	{
		return size() == 1 ? true : false;
	}
	
	public boolean isModified()
	{
		return modified;
	}
	
	public void setModified(boolean modified)
	{
		this.modified = modified;
		this.domain.setModified(modified);
	}
	
	public int row()
	{
		return row;
	}
	
	public int col()
	{
		return col;
	}
	
	public int block()
	{
		return block;
	}
	
	/**
	 * Returns the value currently assigned to the IntVariable.
	 * If the IntVariable is currently unassigned, returns 0
	 * @return 0 if IntVariable is unassigned, Assignment otherwise
	 */
	public Integer getAssignment()
	{
		if (!this.isAssigned())
		{
			return 0;
		}
		else
		{
			return domain.getValues().get(0);
		}
	}
	
	public List<Integer> Values()
	{
		return domain.getValues();
	}
	
	public int size()
	{
		return domain.size();
	}
	
	public Domain getDomain()
	{
		return domain;
	}
	
	public String getName()
	{
		return name;
	}
	
	//===============================================================================
	// Modifiers
	//===============================================================================
	
	/** 
	 * Used for local Search
	 * @param val new assignment
	 */
	public void assignValue(int val)
	{
		setDomain(new Domain(val));
	}
	
	/**
	 * Used for Backtracking. Adds the changed domain to the Trail.
	 * @param d new domain
	 */
	public void updateDomain(Domain d)
	{
		methodModifiesDomain();
		
		if(!domain.equals(d))
		{
			domain = d;
			modified = true;
		}
	}
	
	/**
	 * Performs a change to the domain without changing the Trail.
	 * @param d new domain 
	 */
	public void setDomain(Domain d)
	{
		if(!domain.equals(d))
		{
			domain = d;
			modified = true;
		}
	}
	
	/**
	 * Removes a single value from the domain of V.
	 * @param val value to remove
	 */
	public void removeValueFromDomain(int val) {
		methodModifiesDomain();
		domain.remove(val);
		this.modified=domain.isModified();
	}
	
	//===============================================================================
	// Helper Methods
	//===============================================================================
	
	public void methodModifiesDomain()
	{
		int newSize = size();
		if (oldSize > newSize)
		{
			oldSize = newSize;
		}
			
		trail.push(this);
	}
	
	//===============================================================================
	// Iterator
	//===============================================================================
	
	@Override
	public Iterator<Integer> iterator() {
		return domain.iterator();
	}
	
	//===============================================================================
	// String Representation
	//===============================================================================
	public String toString()
	{
		//prints node stats
		StringBuilder sb = new StringBuilder();
		sb.append(" Name: "+name);
		sb.append("\tdomain: {");
		for(Integer i : domain)
		{
			sb.append(i + ",");	
		}
		sb.deleteCharAt(sb.length()-1);
		sb.append("}");
		return sb.toString();
	}

	
}
